src/simulation_strategies/nonuniform_losses_approximation_strategy.py

```diff
@@ -0,0 +1,83 @@
+    prepare_interferometer_matrix_in_expanded_space_with_first_k_lossless_modes
+from src.quantum_computations_utilities import compute_qft_matrix
+from src.simulation_strategies.lossy_networks_generalized_cliffords_simulation_strategy import \
+    BSPermanentCalculatorInterface, LossyNetworksGeneralizedCliffordsSimulationStrategy
+
+
+class NonuniformLossesApproximationStrategy(LossyNetworksGeneralizedCliffordsSimulationStrategy):
+
+    def __init__(self, bs_permanent_calculator: BSPermanentCalculatorInterface, approximated_modes_number: int) -> None:
+
+        if approximated_modes_number > bs_permanent_calculator.matrix.shape[0]:
+            approximated_modes_number = bs_permanent_calculator.matrix.shape[0]
+        if approximated_modes_number < 0:
+            approximated_modes_number = 0
+
+        self._approximated_modes_number = approximated_modes_number
+        self._modes_transmissivity_values = get_modes_transmissivity_values_from_matrix(bs_permanent_calculator.matrix)
+        self._initial_matrix = prepare_interferometer_matrix_in_expanded_space_with_first_k_lossless_modes\
+            (bs_permanent_calculator.matrix, self._approximated_modes_number)
+        super().__init__(bs_permanent_calculator)
+
+    def simulate(self, input_state: ndarray, samples_number: int = 1) -> List[ndarray]:
+
+        samples = []
+
+        for _ in range(samples_number):
+            lossy_input = self._compute_lossy_input(input_state)
+            self._helper_strategy.set_new_matrix(self._get_matrix_for_approximate_sampling())
+            samples.append(super().simulate(lossy_input)[0])
+
+        return samples
+
+    def _compute_lossy_input(self, input_state: ndarray) -> ndarray:
+
+        if self._approximated_modes_number < 1:
+            return input_state
+
+        lossy_input = deepcopy(input_state)
+
+        binned_input_index = self._approximated_modes_number - 1
+        for i in range(self._approximated_modes_number):
+            if uniform(0, 1) > self._modes_transmissivity_values[i]:
+                lossy_input[binned_input_index] -= 1
+
+        return lossy_input
+
+    def _get_matrix_for_approximate_sampling(self) -> ndarray:
+        qft_matrix = self._get_qft_matrix()
+        random_phases_matrix = self._get_random_phases_matrix()
+        return qft_matrix @ random_phases_matrix @ self._initial_matrix
+
+    def _get_qft_matrix(self):
+        small_qft_matrix = compute_qft_matrix(self._approximated_modes_number)
+        qft_matrix = eye(self._initial_matrix.shape[0], dtype=complex128)
+
+        for i in range(self._approximated_modes_number):
+            for j in range(self._approximated_modes_number):
+                qft_matrix[i][j] = small_qft_matrix[i][j]
+
+        return qft_matrix
+
+    def _get_random_phases_matrix(self) -> ndarray:
+        random_phases_matrix = eye(self._initial_matrix.shape[0], dtype=complex128)
+
+        for i in range(self._approximated_modes_number):
```

Comment:
I believe that this could be replaced with something like:

```
diagonal_vector = numpy.exp(1j * numpy.random.rand(self._approximated_modes_number))
numpy.fill_diagonal(random_phases_matrix, diagonal_vector)
```

Reply:
This probably could work, but I don't really see how it's better. Are there any arguments for this construction or is it just an alternative?

Reply:
It’s just that loops are terribly inefficient in python. If your array sizes go beyond thousands, this could prove a major performance boost.